docs/proposals/20210524-managed-external-etcd.md

```diff
@@ -0,0 +1,570 @@
+* Populating and updating a field that will contain the etcd cluster endpoints (IP addresses of all etcd members). 
+* Creating/managing two Kubernetes Secrets that will contain certs required by the API server to communicate with the etcd cluster
+    * One Secret containing the etcd CA cert, by the name `{cluster.Name}-etcd`
+    * One Secret containing the API server cert-key pair, by the name `{cluster.Name}-apiserver-etcd-client`
+
+###### Required services:
+The etcd provider should install etcd with the version specified by the user for each member.
+
+###### Required fields:
+Each etcd provider should add a new API type to manage the etcd cluster lifecycle, and it should have the following required fields:
+- Spec:
+    - Replicas: Allow only odd numbers, and immutable after creation.
+- Status
+    - CreationComplete: To be set to true once all etcd members pass healthcheck after initial cluster creation.
+    - Endpoints: A comma separated string containing all etcd members endpoints.
+
+###### Contract between etcd provider and control plane provider
+- Control plane providers should check for the presence of the paused annotation (`cluster.x-k8s.io/paused`), and not continue provisioning if it is set. Out of the existing control plane provider, the KubeadmControlPlane controller does that.
+- Control plane providers should check for the presence of `Cluster.Spec.ManagedExternalEtcdRef` field. This check should happen after the control plane is no longer paused, and before the provider starts provisioning the control plane.
+- The control plane provider should "Get" the CR referred by the cluster.spec.ManagedExternalEtcdRef and check its status.Endpoints field. 
+  - This field will be a comma separated list of etcd endpoints. 
+  - It will parse the endpoints from this field and use these endpoints as the external etcd endpoints where applicable. For instance, the KCP controller will read the status.Endpoints field and parse it into a string slice, and then use it as the [external etcd endpoints](https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/api/v1alpha4/kubeadm_types.go#L269).
+- The control plane provider will read the etcd CA cert, and the client cert-key pair from the two Kubernetes Secrets named `{cluster.Name}-apiserver-etcd-client` and `{cluster.Name}-etcd`. The KCP controller currently does that.
+
+###### Contract between etcd provider and infrastructure provider
+- Each etcd member requires port 2379 for client communication and port 2380 for peer communication.
+- In case of the infrastructure providers that use security groups, either the infrastructure provider may add a new security group specifically for etcd members, or document that the user needs to create a security group allowing these two ports and use that for the etcd machines.
+
+![etcd-controller-contract](images/managed-etcd/etcd-controller-contract.png)
+---
+
+
+##### Etcdadm based etcd provider
```

Comment:
If I got this right, the proposed implementation for etcdadm assume a CAPI machine object being created for linking an infrastructure machine to a etcdadmConfig object.

CAPI as of today assumes that each machine hosts K8s node, and the MachineController, The ClusterCacheTracker, MHC and also external components like the autoscaler are relying on this assumption.
How does this works for etcd cluster machines? are we hosting nodes on this machines? how they are represented in the Cluster? how do we prevent workloads to be scheduled on those nodes?

Also, does this have impacts for providers/the image build process? does the images used for etcd machine require different components being pre-installed (etcdadm)? does those images follow the same version schema of control plane/worker machines